Blog about formal syntax work (#50)
openwebdocs/owd-website: pages/content/posts/improve-formal-syntax/index.md

---
author: "Will Bamberg"

> title: "Improving the CSS formal syntax"

Reviewer: ```suggestion
title: "Improving MDN's CSS formal syntax"
```

> date: "2022-10-06"

Reviewer: How about publishing this on Tuesday?

```suggestion
date: "2022-10-11"
```

tags: ["projects"]
ShowToc: false
ShowBreadCrumbs: false

> excerpt: "How we made the CSS formal syntax more readable and reliable, with the help of some great open source software packages."

Reviewer: ```suggestion
excerpt: "How we made MDN's CSS formal syntax more readable and reliable, with the help of some great open source software packages."
```

---

> If you've spent much time with our CSS docs, you've probably seen the "Formal syntax" section in the CSS property reference pages. When we ask developers what they think of this part of the docs, most of them say something like "I have no idea what this means". This used to be my reaction too, but now I think it can be extremely useful. CSS properties can have very complex syntax, and explaining it in prose can be really hard. Once you've learned a few rules, the formal syntax gives you a very precise description of the syntax in a very concise manner.

Reviewer: ```suggestion
If you've spent much time with MDN's CSS docs, you've probably seen the "Formal syntax" section in the CSS property reference pages (see for example the [`background` CSS property](https://developer.mozilla.org/en-US/docs/Web/CSS/background#formal_syntax)). When we ask developers what they think of this part of the docs, most of them say something like "I have no idea what this means". This used to be my reaction too, but now I think it can be extremely useful. CSS properties can have very complex syntax, and explaining it in prose can be really hard. Once you've learned a few rules, the formal syntax gives you a very precise description of the syntax in a very concise manner.
```

Author: ```suggestion
If you've spent much time with the CSS docs on MDN, you've probably seen the ["Formal syntax" section](https://developer.mozilla.org/en-US/docs/Web/CSS/background#formal_syntax) in the CSS property reference pages. When we ask developers what they think of this part of the docs, most of them say something like "I have no idea what this means". This used to be my reaction too, but now I think it can be extremely useful. CSS properties can have very complex syntax, and explaining it in prose can be really hard. Once you've learned a few rules, the formal syntax gives you a very precise description of the syntax in a very concise manner.
```

is that OK? Avoiding the possessive "MDN's" which is a bit ugly, and making the link part more flowy.

> If you've spent much time with the CSS docs on MDN, you've probably seen the ["Formal syntax" section](https://developer.mozilla.org/en-US/docs/Web/CSS/background#formal_syntax) in the CSS property reference pages. When we ask developers what they think of this part of the docs, most of them say something like "I have no idea what this means". This used to be my reaction too, but now I think it can be extremely useful. CSS properties can have very complex syntax, and explaining it in prose can be really hard. Once you've learned a few rules, the formal syntax gives you a very precise description of the syntax in a very concise manner.

Author: ```suggestion
If you've spent much time with the CSS docs on MDN, you've probably seen the ["Formal syntax" section](https://developer.mozilla.org/en-US/docs/Web/CSS/background#formal_syntax) in the CSS property reference pages. When we ask developers what they think of this part of the docs, most of them say something like "I have no idea what this means". This used to be my reaction too, but now I think it can be extremely useful. CSS properties can have complex syntax, and explaining it in prose can be really hard. Once you've learned a few rules, the formal syntax gives you a very precise description of the syntax in a very concise manner.
```

> Until recently, though, there were two big problems with the way we represented formal syntax.

Reviewer: ```suggestion
Until recently, though, there were two big problems with the way MDN represented formal syntax.
```

The first problem was presentation: the macro that rendered the formal syntax as HTML tended to render it as a wall of text, making it hard to read and scarier than strictly necessary:

![Screenshot of old formal syntax rendering for the background property, showing it as a wall of text](images/formal-syntax-background-before.png)

> The second problem was about data quality. Although the formal syntax should match the syntax in the specifications, we maintained our own copy of the data in a separate repository, [mdn/data](https://github.com/mdn/data). We often weren’t able to keep this up to date with the specification, and as a result, users of MDN would see inaccurate or outdated syntax.

Reviewer: ```suggestion
The second problem was about data quality. Although the formal syntax should match the syntax in the specifications, MDN maintained its own copy of the data in a separate repository, [mdn/data](https://github.com/mdn/data). We often weren’t able to keep this up to date with the specification, and as a result, users of MDN would see inaccurate or outdated syntax.
```
